File: python/lexer/lexer.py
from lexer.my_token import TokenType, keywords, token_dict, Token
from dataclasses import dataclass


@dataclass
class Lexer:
    input: str
    position: int = 0
    readPosition: int = 0
    ch: str = ""
    stop: int = 1

    def __init__(self, input: str) -> None:
        self.input = input
        self.readChar()

    def peekChar(self):
        if self.readPosition >= len(self.input):
            return 0
        else:
            return self.input[self.readPosition]

    def readChar(self):
        if self.readPosition >= len(self.input):
            self.stop = 0
            self.ch = ""
        else:
            self.ch = self.input[self.readPosition]
        self.position = self.readPosition
        self.readPosition += 1

    def skipWhitespace(self):
        while self.ch.isspace():
            self.readChar()

    def nextToken(self) -> Token:
        self.skipWhitespace()
        if self.stop == 0:
            tok = Token(TokenType.EOF, "")
        elif token_dict.get(self.ch, False):
            tok = Token(token_dict[self.ch], self.ch)
        elif self.ch == '=':
            if self.peekChar() == '=':
                self.readChar()
                tok = Token(TokenType.EQ, '==')
            else:
                tok = Token(TokenType.ASSIGN, self.ch)
        elif self.ch == '!':
            if self.peekChar() == '=':
                self.readChar()
                tok = Token(TokenType.NOT_EQ, '!=')
            else:
                tok = Token(TokenType.BANG, self.ch)

        elif self.ch.isalpha():
            literal = self.readIdentifier()
            return Token(lookUpIdent(literal), literal)
        elif self.ch.isnumeric():
            tok_type = TokenType.INT
            literal = self.readNumber()
            return Token(tok_type, literal)
        else:
            tok = Token(TokenType.ILLEGAL, self.ch)
        self.readChar()
        return tok

    def readIdentifier(self):
        position = self.position

        while isLetter(self.ch):
            self.readChar()
        return self.input[position:self.position]

    def readNumber(self):
        position = self.position
        while isDigit(self.ch):
            self.readChar()

        return self.input[position:self.position]


def isLetter(ch: str):
    return 'a' <= ch <= 'z' or 'A' <= ch <= 'Z' or ch == '_'


def isDigit(ch: str):
    return '0' <= ch <= '9'


def lookUpIdent(ident) -> TokenType:
    return keywords[ident] if ident in keywords else TokenType.IDENT

File: dart/lib/src/lexer/lexer.dart
import 'utils.dart';
import 'token.dart';

export 'token.dart';

const $EOF = "\u0000";

class Lexer {
  int _pos = 0;
  int _readPos = 0;
  String ch = $EOF;

  final String input;

  Lexer(this.input) {
    _readChar();
  }

  Token nextToken() {
    _skipWhitespace();

    final ch = this.ch;
    final nextCh = _peekChar();
    final tok = switch ((ch, nextCh)) {
      ("=", "=") => _skipNextAs(Token.equal()),
      ("=", _) => Token.assign(),
      ("!", "=") => _skipNextAs(Token.notEqual()),
      ("!", _) => Token.bang(),
      ("+", _) => Token.plus(),
      ("-", _) => Token.minus(),
      ("/", _) => Token.slash(),
      ("*", _) => Token.asterisk(),
      ("<", "=") => _skipNextAs(Token.lte()),
      (">", "=") => _skipNextAs(Token.gte()),
      ("<", _) => Token.lt(),
      (">", _) => Token.gt(),
      (",", _) => Token.comma(),
      (";", _) => Token.semicolon(),
      ("(", _) => Token.lParen(),
      (")", _) => Token.rParen(),
      ("{", _) => Token.lSquirly(),
      ("}", _) => Token.rSquirly(),
      ($EOF, _) => Token.eof(),
      (String c, _) when c.isLetter() => _readIdentifier(),
      (String c, _) when c.isNumber() => _readNumber(),
      _ => Token(TokenType.illegal, ch)
    };

    _readChar();
    return tok;
  }

  Token _lookupIdent(String ident) {
    return switch (ident) {
      "let" => Token.let(),
      "fn" => Token.fn(),
      "true" => Token.true_(),
      "false" => Token.false_(),
      "if" => Token.if_(),
      "else" => Token.else_(),
      "return" => Token.return_(),
      _ => Token.ident(ident),
    };
  }

  void _skipWhitespace() {
    while (ch.isWhitespace()) {
      _readChar();
    }
  }

  void _readChar() {
    ch = switch (_readPos) {
      int p when p >= input.length => $EOF,
      _ => input[_readPos],
    };

    _pos = _readPos;
    _readPos += 1;
  }

  String _peekChar() {
    return switch (_readPos) {
      int p when p >= input.length => $EOF,
      _ => input[_readPos],
    };
  }

  void _rewind() {
    _pos -= 1;
    _readPos -= 1;
    ch = input[_pos];
  }

  T _skipNextAs<T>(T t) {
    _readChar();
    return t;
  }

  Token _readIdentifier() {
    final pos = _pos;
    while (ch.isLetter()) {
      _readChar();
    }
    final id = input.substring(pos, _pos);
    _rewind();
    return _lookupIdent(id);
  }

  Token _readNumber() {
    final pos = _pos;
    while (ch.isNumber()) {
      _readChar();
    }
    final number = input.substring(pos, _pos);
    _rewind();
    return Token.int(number);
  }

  Iterable<Token> iter() sync* {
    var tok = nextToken();
    while (tok.type != TokenType.eof) {
      yield tok;
      tok = nextToken();
    }
  }
}
